### PR TITLE
Use non-global RNGs

### PR DESCRIPTION
**Pull request by devmotion**
Fixes https://github.com/TuringLang/Turing.jl/issues/1260.

### CODE REVIEW

File: test/inference/Inference.jl [hunk 1]
```diff
@@ -12,13 +12,25 @@ include(dir*"/test/test_utils/AllUtils.jl")
...
+
+                samplers = (HMC(0.1, 7),
+                            PG(10),
+                            IS(),
+                            MH(),
+                            Gibbs(PG(3, :s), HMC(0.4, 8, :m)),
+                            Gibbs(HMC(0.1, 5, :s), ESS(:m)))
+                for sampler in samplers
+                    Random.seed!(5)
+                    chain1 = sample(model, sampler, MCMCThreads(), 1000, 4)
```
Comment (yebai) [inline test/inference/Inference.jl:26]: How threads rng are initialised? Are they sharing the same random seed?

Comment (devmotion) [reply test/inference/Inference.jl:26]: That is handled in AbstractMCMC: https://github.com/TuringLang/AbstractMCMC.jl/blob/509b5cf3fe24c6a8809ba6ca98f42ae8165e12e2/src/sample.jl#L215-L216
We sample separate seeds for every chain, using the provided RNG, and create separate copies of the RNG for every thread. When sampling the `i`th chain, the RNG copy of the active thread is seeded with the `i`th seed. Hence the results can be reproduced even with different numbers of threads. The same happens when sampling with multiple processes.

Comment (yebai) [reply test/inference/Inference.jl:26]: thanks for the clarification!

**Comment by yebai** [comment]
Ready to merge once the tests pass.

**Comment by torfjelde** [comment]
Should we add `predict(rng::AbstractRNG, ...)` as we spoke about @devmotion in this PR or should I make another one after this has gotten merged?:)

**Comment by devmotion** [comment]
The PR is based on Turing master before `predict` and `MLE` and `MAP` were added, so if we want to fix `predict` we would have to merge master first (or rebase). I don't mind if it's addressed in this PR or in a separate one, but I guess we should also make sure that VI uses user-provided RNGs.

**Comment by cpfiffer** [comment]
It'll be cleaner if done separately.

**Pull request merged**